drivers/nvdimm/pmem.c

```diff
@@ -45,11 +65,352 @@ static struct nd_region *to_region(struct pmem_device *pmem)
 	return to_nd_region(to_dev(pmem)->parent);
 }
 
+static bool pmem_dma_feq(struct dma_chan *chan, void *data)
+{
+	return dev_to_node(chan->device->dev) == (long)data;
+}
+
+static bool pmem_dma_fne(struct dma_chan *chan, void *data)
+{
+	return dev_to_node(chan->device->dev) != (long)data;
+}
+
+static void pmem_dma_init(struct pmem_device *pmem)
+{
+	struct device *dev = to_dev(pmem);
+	dma_cap_mask_t dma_mask;
+
+	dma_cap_zero(dma_mask);
+	dma_cap_set(DMA_MEMCPY, dma_mask);
+
+	/*
+	 * Prefer to allocate NUMA-local DMA channel for remote PMEM writes.
+	 * Remote writes are slower than local and need all boost we can give.
+	 */
+	pmem->rdma_chan = dma_request_channel(dma_mask, pmem_dma_feq,
+	    (void *)(long)pmem->rnode);
+	if (!pmem->rdma_chan)
+		pmem->rdma_chan = dma_request_channel(dma_mask, NULL, NULL);
+	if (!pmem->rdma_chan) {
+		dev_info(dev, "Unable to allocate DMA channel\n");
+		return;
+	}
+
+	/*
+	 * If local PMEM is in different NUMA-node, try to allocate DMA there.
+	 * If it is in the same NUMA-node, then DMA allocation from some other
+	 * allows to avoid DMA engine bottleneck and shows better throughput.
+	 */
+	if (pmem->node != dev_to_node(pmem->rdma_chan->device->dev)) {
+		pmem->dma_chan = dma_request_channel(dma_mask, pmem_dma_feq,
+		    (void *)(long)pmem->node);
+	} else {
+		pmem->dma_chan = dma_request_channel(dma_mask, pmem_dma_fne,
+		    (void *)(long)dev_to_node(pmem->rdma_chan->device->dev));
+	}
```

Comment:
Same observation here. Conditionals at 111 and 113 look identical... unless my aging eyes are playing games on me 😝

Reply:
👍🏼